### DISCUSSION TITLE
Why is word match not working?

### DISCUSSION THREAD

**Discussion by rew1nter** [BurntSushi/ripgrep#2775] [Q&A] [answered]
Isn't the behavior supposed to be the same as grep here?

![image](https://github.com/BurntSushi/ripgrep/assets/64508791/d1e4a72d-6764-4c3e-86de-34a07098a960)

**Comment by BurntSushi** [comment, marked as answer]
This has nothing to do with the `--w/--word-regexp` flag. This is due to the fact that grep uses [basic regular expressions](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html) (BREs) by default. BREs have different escaping rules and different rules around when `*` is "special" or not. In this case, since it appears the beginning of the regex, it isn't "special" and matched literally. If you use extended regular expressions (EREs) instead, then the `*` is treated specially:

```
$ wmctrl -d | grep -E '*'
grep: warning: * at start of expression
0  - DG: 2560x1600  VP: N/A  WA: N/A  1
1  - DG: 2560x1600  VP: N/A  WA: N/A  2
2  - DG: 2560x1600  VP: N/A  WA: N/A  3
3  - DG: 2560x1600  VP: N/A  WA: N/A  4
4  * DG: 2560x1600  VP: N/A  WA: N/A  mail
5  - DG: 2560x1600  VP: N/A  WA: N/A  browser
6  - DG: 2560x1600  VP: N/A  WA: N/A  text
7  - DG: 2560x1600  VP: N/A  WA: N/A  chat
8  - DG: 2560x1600  VP: N/A  WA: N/A  s1
9  - DG: 2560x1600  VP: N/A  WA: N/A  w1
10 - DG: 2560x1600  VP: N/A  WA: N/A  doc
11 - DG: 2560x1600  VP: N/A  WA: N/A  a
12 - DG: 2560x1600  VP: N/A  WA: N/A  k
13 - DG: 2560x1600  VP: N/A  WA: N/A  ecma
14 - DG: 2560x1600  VP: N/A  WA: N/A  fwd
15 - DG: 2560x1600  VP: N/A  WA: N/A  play
```

In this case, `*` is probably being treated as being applied to an empty sub-expression and thus matching everything. ripgrep instead recognizes the expression as invalid and rejects it.

Apparently POSIX says that the regex `*` produces "undefined results," so I guess grep can do whatever it wants here. This is the insanity of POSIX. It defines a miniature language (regexes) to be used by end users, and specifically leaves huge "undefined" holes in the syntax.

> supposed to be the same as grep

In the general case, no, it isn't: https://github.com/BurntSushi/ripgrep/blob/master/FAQ.md#posix4ever
